### PR TITLE
tests: gadget 0: added test for unaligned buffer read/writes.

### PR DESCRIPTION
**Pull request by fenugrec**
[STATUS - this is ready to merge to the best of my knowledge]

This adds a test to the gadget0 framework to test unaligned vs aligned buffer access in the usb stack, and the corresponding fix.
Without the fix, this fails on stm32F072, which is expected (all F0 / L0 mcus should fail) but not normal. Note that once the unaligned test fails, the target is in HardFault and needs to be reset.

See issues #401, #460.

I am only able to test on a F072 target, it would be nice if someone could test on something other than F0* / L0*, for example F1 which should not exhibit the issue.

The fix is copied from Kuldeep's tree; I haven't modified the st_usbfs_copy_to_pm() function however,  since it's already a bytewise copy.

### CODE REVIEW

File: tests/gadget-zero/test_gadget0.py [hunk 1]
```diff
@@ -318,3 +318,56 @@ def test_mean_limits(self):
...
+        self.assertEqual(len(data), read_size, "Should have read as much as we asked for")
+
+    def test_aligned(self):
+        self.set_aligned()
+        self.do_readwrite()
+
+    def test_unaligned(self):
+        self.set_unaligned()
+        self.do_readwrite()
+        self.set_aligned()
```
Comment (karlp) [inline tests/gadget-zero/test_gadget0.py]: shouldn't need this last line

File: tests/gadget-zero/usb-gadget0.c [hunk 2]
```diff
@@ -159,7 +162,7 @@ static const struct usb_config_descriptor config[] = {
 	}
 };
 
-static char serial[] = "0123456789.0123456789.0123456789";
+static char serial[] = "stm32f072disco";
```
Comment (karlp) [inline tests/gadget-zero/usb-gadget0.c]: why did you need to do this?

File: tests/gadget-zero/usb-gadget0.c [hunk 3]
```diff
@@ -181,37 +184,60 @@ static struct {
...
 static void gadget0_ss_in_cb(usbd_device *usbd_dev, uint8_t ep)
 {
 	(void) usbd_dev;
 	trace_send_blocking8(0, 'I');
-	uint8_t buf[64];
+	uint16_t buf[BULK_EP_MAXPACKET / 2 + 1];
+	uint8_t *src;
+
+	if (test_unaligned) {
+		src = ((uint8_t *) buf) + 1;
```
Comment (karlp) [inline tests/gadget-zero/usb-gadget0.c]: how do you know that it was aligned before? is that because you used a uint16_t buf that must get aligned?  Why not just use __attribute__(aligned(4)) on the uint8_t buffer, and +1 there?  more obvious what's going on?

File: tests/gadget-zero/usb-gadget0.c [hunk 4]
```diff
@@ -279,6 +311,7 @@ static void gadget0_set_config(usbd_device *usbd_dev, uint16_t wValue)
 	ER_DPRINTF("set cfg %d\n", wValue);
 	switch (wValue) {
 	case GZ_CFG_SOURCESINK:
+		test_unaligned = 0;
```
Comment (karlp) [inline tests/gadget-zero/usb-gadget0.c]: can this just go into the state structure? much like the pattern_counter?

File: tests/gadget-zero/usb-gadget0.c [hunk 5]
```diff
@@ -181,37 +184,60 @@ static struct {
 	.pattern_counter = 0,
 };
 
+/* aligned / unaligned test flag.
+ * Using a bool type for this instead, in some cases, can cause gcc to place other critical buffers
```
Comment (karlp) [inline tests/gadget-zero/usb-gadget0.c]: what?

File: tests/gadget-zero/usb-gadget0.c (hunk 2 above)
Comment (fenugrec) [inline tests/gadget-zero/usb-gadget0.c]: didn't cross my mind; of course that line won't be part of the final PR.

File: tests/gadget-zero/usb-gadget0.c (hunk 3 above)
Comment (fenugrec) [inline tests/gadget-zero/usb-gadget0.c]: I was avoiding gcc-specific directives, and uint16_t should guarantee alignment on M0/M0+... but I guess it's not that much clearer this way. I think I'll change it to aligned(2)

File: tests/gadget-zero/usb-gadget0.c (hunk 4 above)
Comment (fenugrec) [inline tests/gadget-zero/usb-gadget0.c]: Certainly, I wasn't sure if I could do that or if the state structure was reserved for the linux-gadget-compatible functionality.

File: tests/gadget-zero/usb-gadget0.c (hunk 5 above)
Comment (fenugrec) [inline tests/gadget-zero/usb-gadget0.c]: That's what caused my problems. I had a global bool, that gcc allocated as a uint8 , and in certain conditions other uint8_t arrays were allocated right after, with no padding, causing an unaligned control buffer and associated issues.

File: tests/gadget-zero/usb-gadget0.c (hunk 4 above)
Comment (karlp) [inline tests/gadget-zero/usb-gadget0.c]: no, not at all, I'm only aiming for functional equiv, not code equiv

File: tests/gadget-zero/usb-gadget0.c [hunk 6]
```diff
@@ -176,42 +179,60 @@ static usbd_device *our_dev;
 static struct {
 	uint8_t pattern;
 	int pattern_counter;
+	int test_unaligned;	/* If 0 (default), use 16-bit aligned buffers. This should not be declared as bool */
```
Comment (karlp) [inline tests/gadget-zero/usb-gadget0.c]: comment no longer relevant as you're using attributes on buffer

**Comment by karlp** [comment]
:+1:

**Comment by karlp** [comment]
Thanks for this! merged with some splits and rewords
